engine/source_sym_engine.ml

```diff
@@ -0,0 +1,184 @@
+open Ast
+
+type constraint_tree =
+  | Leaf of pi list * source_expr
+  | Node of pi * constraint_tree list
+  | Conjunction of constraint_tree list
+  | Root of constraint_tree list
+and
+  pi = { idx: int; op: piop }
+and
+  piop =
+  | IsK of string (* Variant *)
+  | NotK of string (* All variants except k *)
+  | IsTrue 
+  | IsFalse 
+  | IsTuple 
+  | NotTuple 
+  | IsNil
+  | NotNil
+  | IsCons
+  | NotCons
+  | EqInt of int
+  | NeqInt of int
+  | EqString of string
+  | NeqString of string
+
+let print_result stree =
+  let bprintf = Printf.bprintf
+  in
+  let bprint_piop buf piop = match piop with
+    | EqInt i -> bprintf buf "EqInt %d" i
+    | NeqInt i -> bprintf buf "NeqInt %d" i
+    | IsK s -> bprintf buf "IsK %s" s
+    | NotK s -> bprintf buf "NotK %s" s
+    | IsTrue -> bprintf buf "IsTrue"
+    | IsFalse -> bprintf buf "IsFalse"
+    | IsTuple -> bprintf buf "IsTuple"
+    | NotTuple -> bprintf buf "NotTuple"
+    | IsNil -> bprintf buf "IsNil"
+    | NotNil -> bprintf buf "NotNil"
+    | IsCons -> bprintf buf "IsCons"
+    | NotCons -> bprintf buf "NotCons"
+    | EqString s -> bprintf buf "EqString %s" s
+    | NeqString s -> bprintf buf "NeqString %s" s
+  in
+  let bprint_pi buf pi =
+    bprintf buf "{idx=%d; %a}"
+      pi.idx
+      bprint_piop pi.op
+  in
+  let bprint_source_expr buf = function SBlackbox s -> bprintf buf "%s" s in
+  let rec bprint_list ~sep bprint buf = function
+    | [] -> ()
+    | [x] -> bprint buf x
+    | x :: xs ->
+       bprintf buf "%a%t%a"
+         bprint x
+         sep
+         (bprint_list ~sep bprint) xs in
```

Comment:
Note: we could have an `Utils` module with auxiliary functions used in several modules (here pretty-printing helpers), instead of code duplication.

Reply:
Yes, actually I was thinking of a small refactoring of the pretty printing functions in the target_sym_exec module, but that may be the goal of a new PR